org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/UnitTestSupport.java

```diff
@@ -0,0 +1,776 @@
+        } catch(CoreException ex) {
+          log.error(ex.getMessage(), ex);
+        }
+      }
+
+      // executing prerequisites mojo (mainly for properties)
+      for(MojoExecution execution : executions) {
+        try {
+          cache.executeMojo(facade, execution, monitor);
+        } catch(MojoExecutionException | MojoFailureException ex) {
+          log.error(ex.getMessage(), ex);
+        }
+      }
+
+    }
+
+    /**
+     * Parse the custom execution ids
+     * 
+     * @param customExecutions the custom executions
+     * @return the list of execution ids
+     */
+    private List<ExecutionId> parseExecutionIds(String customExecutions) {
+      List<ExecutionId> executionIds = new ArrayList<>();
+      String[] customExecutionsArray = customExecutions.split(",");
+      for(String customExecution : customExecutionsArray) {
+        String[] customExecutionArray = customExecution.split(":");
+        if(customExecutionArray.length == 3) {
+          executionIds.add(new ExecutionId(customExecutionArray[0], customExecutionArray[1], customExecutionArray[2]));
+        }
+      }
+      return executionIds;
+    }
+
+    /**
+     * Check if the test is handled by the plugin
+     * 
+     * @param facade the maven project facade
+     * @param monitor the progress monitor
+     * @param execution the plugin execution
+     * @param testClass the test class
+     * @return true if the test is handled by the plugin
+     */
+
+    private boolean isTestHandledByPlugin(IMavenProjectFacade facade, IProgressMonitor monitor, MojoExecution execution,
+        String testClass) {
+
+      String testFile = testClass.replace(".", "/") + ".class";
+      // get a configured mojo instance 
+      Mojo mojo = cache.getMojoInstance(facade, execution, monitor);
+
+      // get an instance of org.apache.maven.surefire.api.testset.TestListResolver directly from the plugin instance
+      Optional<Object> testResolverInstance = Optional.of(mojo)
+          .map(o -> findMethod(o, GET_INCLUDED_AND_EXCLUDED_TESTS_METHOD)).map(om -> om.orElseThrow()).map(m -> {
+            m.setAccessible(true);
+            return m;
+          }).map(m -> uncheckedInvoke(m, mojo));
+
+      // check if the test is handled by the plugin
+      Boolean isTestHandled = testResolverInstance.filter(Objects::nonNull)
+          .map(o -> findMethod(o, SHOULD_RUN_METHOD, String.class, String.class)).map(om -> om.orElseThrow()).map(m -> {
+            m.setAccessible(true);
+            return m;
+          }).map(m -> uncheckedInvoke(m, testResolverInstance.get(), testFile, "")).map(Boolean.class::cast)
+          .orElse(false);
+
+      return isTestHandled;
+    }
+
+    /**
+     * Lookup method in class and ancestors
+     * 
+     * @param instance the instance on which we search the method
+     * @return an optional method
+     */
+    private Optional<Method> findMethod(Object instance, String methodName, Class<?>... parameters) {
```

Comment:
```suggestion
    private Method findMethod(Object instance, String methodName, Class<?>... parameters) {
```
If you return the Method (or null) directly you can save the intermediate `orElseThrow()` in the caller and terminate the Optional chain earlier and can continue with plain references.

Reply:
ok